Process Eye0 crashed with trace: AttributeError: 'types.SimpleNamespace' object has no attribute 'zmq_ctx'

I created a video backend to stream a video from raspberry to my laptop where pupil capture software is running. The hmd_streaming is then started via a notification. My video stream server behaves like hmd_streaming expects it. 
Here is the code: https://github.com/Lifestohack/pupil-video-backend/blob/master/video_backend.py#L60
I am able to start the world view via the notification. But eye0 and eye1 cannot be started. If I start my video backend server I get the error on the pupil capture software that says: Process Eye0 crashed with trace. 

Looking at the capture.log following Traceback is found:

```
- eye0 - [ERROR] launchables.eye: Process Eye0 crashed with trace:
Traceback (most recent call last):
  File "launchables/eye.py", line 654, in eye
  File "shared_modules/plugin.py", line 398, in add
  File "shared_modules/video_capture/hmd_streaming.py", line 50, in __init__
AttributeError: 'types.SimpleNamespace' object has no attribute 'zmq_ctx'
```

Reproduction steps:

1. Start pupil capture software.
2. Click on "Detect eye 0" on pupil capture software. Small window pops up where eye0 video feed supposed to be shown.
3. Started pupil video backend.

But pupil detection eye 0 window closes automatically. And pupil capture shows the error: Process Eye0 crashed with trace.

```
2020-11-12 13:32:15,630 - world - [DEBUG] network_api.controller.pupil_remote_controller: Request: 'PUB_PORT', Response: '44017'
2020-11-12 13:32:15,634 - world - [DEBUG] network_api.controller.pupil_remote_controller: Request: 'notify.start_eye_plugin', Response: 'Message forwarded.'
2020-11-12 13:32:15,656 - eye0 - [DEBUG] plugin: Replacing Video Source due to 'by_base_class' uniqueness
2020-11-12 13:32:15,656 - eye0 - [DEBUG] plugin: Unloaded Plugin: <video_capture.uvc_backend.UVC_Source object at 0x7f0e638784a8>
2020-11-12 13:32:15,657 - eye0 - [ERROR] launchables.eye: Process Eye0 crashed with trace:
Traceback (most recent call last):
  File "launchables/eye.py", line 654, in eye
  File "shared_modules/plugin.py", line 398, in add
  File "shared_modules/video_capture/hmd_streaming.py", line 50, in __init__
AttributeError: 'types.SimpleNamespace' object has no attribute 'zmq_ctx'
```

--- a/pupil_src/launchables/eye.py
+++ b/pupil_src/launchables/eye.py
@@ -196,7 +196,11 @@ def eye(
         g_pool.timebase = timebase
         g_pool.camera_render_size = None
 
+        g_pool.zmq_ctx = zmq_ctx
         g_pool.ipc_pub = ipc_socket
+        g_pool.ipc_pub_url = ipc_pub_url
+        g_pool.ipc_sub_url = ipc_sub_url
+        g_pool.ipc_push_url = ipc_push_url
 
         def get_timestamp():
             return get_time_monotonic() - g_pool.timebase.value

--- a/pupil_src/shared_modules/video_capture/hmd_streaming.py
+++ b/pupil_src/shared_modules/video_capture/hmd_streaming.py
@@ -78,10 +78,6 @@ class BGRFrame(Uint8BufferFrame):
 
 class RGBFrame(BGRFrame):
     @property
-    def depth(self) -> int:
-        return 3
-
-    @property
     def bgr(self) -> np.ndarray:
         try:
             return self._bgr
@@ -98,7 +94,7 @@ class RGBFrame(BGRFrame):
             return self._gray
 
 
-class GrayFrame(BGRFrame):
+class GrayFrame(Uint8BufferFrame):
     @property
     def depth(self) -> int:
         return 1

--- a/pupil_src/shared_modules/video_capture/hmd_streaming.py
+++ b/pupil_src/shared_modules/video_capture/hmd_streaming.py
@@ -46,15 +46,21 @@ FRAME_CLASS_BY_FORMAT = {"rgb": RGBFrame}
 class HMD_Streaming_Source(Base_Source):
     name = "HMD Streaming"
 
-    def __init__(self, g_pool, *args, **kwargs):
+    def __init__(self, g_pool, topics=("hmd_streaming.world",), *args, **kwargs):
         super().__init__(g_pool, *args, **kwargs)
         self.fps = 30
         self.projection_matrix = None
+        self.__topics = topics
         self.frame_sub = zmq_tools.Msg_Receiver(
             self.g_pool.zmq_ctx,
             self.g_pool.ipc_sub_url,
-            topics=("hmd_streaming.world",),
+            topics=self.__topics,
         )
+
+    def get_init_dict(self):
+        init_dict = super().get_init_dict()
+        init_dict["topics"] = self.__topics
+        return init_dict
 
     def cleanup(self):
         self.frame_sub = None

--- a/pupil_src/shared_modules/video_capture/hmd_streaming.py
+++ b/pupil_src/shared_modules/video_capture/hmd_streaming.py
@@ -35,6 +35,10 @@ class RGBFrame:
         self.yuv_buffer = None
         self.jpeg_buffer = None
 
+    @property
+    def gray(self):
+        return np.mean(self.bgr, axis=-1)
+
 
 FRAME_CLASS_BY_FORMAT = {"rgb": RGBFrame}
 

--- a/pupil_src/shared_modules/video_capture/hmd_streaming.py
+++ b/pupil_src/shared_modules/video_capture/hmd_streaming.py
@@ -9,6 +9,7 @@ See COPYING and COPYING.LESSER for license details.
 ---------------------------------------------------------------------------~(*)
 """
 
+import abc
 import logging
 
 import numpy as np
@@ -21,12 +22,10 @@ from video_capture.base_backend import Base_Source
 logger = logging.getLogger(__name__)
 
 
-class RGBFrame:
+class Uint8BufferFrame(abc.ABC):
     def __init__(self, buffer, timestamp, index, width, height):
-
-        rgb = np.fromstring(buffer, dtype=np.uint8).reshape(height, width, 3)
-        self.bgr = np.ascontiguousarray(np.flip(rgb, (0, 2)))
-        self.img = self.bgr
+        #
+        self._buffer = self.interpret_buffer(buffer, width, height)
         self.timestamp = timestamp
         self.index = index
         self.width = width
@@ -35,12 +34,95 @@ class RGBFrame:
         self.yuv_buffer = None
         self.jpeg_buffer = None
 
+    def interpret_buffer(self, buffer, width, height) -> np.ndarray:
+        return np.fromstring(buffer, dtype=np.uint8).reshape(height, width, self.depth)
+
+    @property
+    @abc.abstractmethod
+    def depth(self) -> int:
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def gray(self) -> np.ndarray:  # dtype uint8, shape (height, width)
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def bgr(self) -> np.ndarray:
+        # dtype uint8, shape (height, width, 3), memory needs to be allocated contiguous
+        raise NotImplementedError
+
+    @property
+    def img(self) -> np.ndarray:  # equivalent for bgr; kept for legacy reasons
+        return self.bgr
+
+
+class BGRFrame(Uint8BufferFrame):
+    @property
+    def depth(self) -> int:
+        return 3
+
+    @property
+    def bgr(self) -> np.ndarray:
+        return self._buffer
+
     @property
     def gray(self):
-        return np.mean(self.bgr, axis=-1).astype(self.bgr.dtype)
+        try:
+            return self._gray
+        except AttributeError:
+            self._gray = np.mean(self._buffer, axis=-1).astype(self._buffer.dtype)
+            return self._gray
 
 
-FRAME_CLASS_BY_FORMAT = {"rgb": RGBFrame}
+class RGBFrame(BGRFrame):
+    @property
+    def depth(self) -> int:
+        return 3
+
+    @property
+    def bgr(self) -> np.ndarray:
+        try:
+            return self._bgr
+        except AttributeError:
+            self._bgr = np.ascontiguousarray(np.flip(self._buffer, (0, 2)))
+            return self._bgr
+
+    @property
+    def gray(self):
+        try:
+            return self._gray
+        except AttributeError:
+            self._gray = np.mean(self._buffer, axis=-1).astype(self._buffer.dtype)
+            return self._gray
+
+
+class GrayFrame(BGRFrame):
+    @property
+    def depth(self) -> int:
+        return 1
+
+    @property
+    def bgr(self) -> np.ndarray:
+        try:
+            return self._bgr
+        except AttributeError:
+            self._bgr = np.ascontiguousarray(np.dstack([self._buffer] * 3))
+            return self._bgr
+
+    @property
+    def gray(self):
+        return self._buffer
+
+    def interpret_buffer(self, buffer, width, height) -> np.ndarray:
+        buffer = super().interpret_buffer(buffer, width, height)
+        # since this will be our gray buffer, we need to get rid of our third dimension
+        buffer.shape = height, width
+        return buffer
+
+
+FRAME_CLASS_BY_FORMAT = {"rgb": RGBFrame, "bgr": BGRFrame, "gray": GrayFrame}
 
 
 class HMD_Streaming_Source(Base_Source):

--- a/pupil_src/shared_modules/video_capture/hmd_streaming.py
+++ b/pupil_src/shared_modules/video_capture/hmd_streaming.py
@@ -37,7 +37,7 @@ class RGBFrame:
 
     @property
     def gray(self):
-        return np.mean(self.bgr, axis=-1)
+        return np.mean(self.bgr, axis=-1).astype(self.bgr.dtype)
 
 
 FRAME_CLASS_BY_FORMAT = {"rgb": RGBFrame}

